### PR TITLE
Support for a generic filelist-based database

### PR DESCRIPTION
**Pull request by pkorshunov**
Integrated into pyannote.database this implementation by @hbredin: https://gist.github.com/hbredin/5fb26bf2b4ea3b38e1df34fd05d06911

### CODE REVIEW

**Comment by hbredin** [comment]
Thanks @pkorshunov.

Could you please give me permission to push to your repo so that we can work together on this?
See https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by hbredin** [comment]
`pyannote.database` now has three different configuration files:
- `~/.pyannote/db.yml` is used by `FileFinder` to find files.
- `~/.pyannote/meta.yml` is used to define meta-protocol
- `~/.pyannote/protocols.yml` is used for this generic database interface.

I think this is getting too complex. We should merge everything into a unique `~/.pyannote/database.yml` whose structure has yet to be defined...

File: pyannote/database/__init__.py [hunk 1]
```diff
@@ -36,10 +36,12 @@
 
 from .database import Database
 from .database import PyannoteDatabaseException
+from .filelist import add_filelist_databases
```
Comment (hbredin) [inline pyannote/database/__init__.py]: I'd rather use `custom` than `filelist`.

File: pyannote/database/__init__.py [hunk 2]
```diff
@@ -56,6 +58,9 @@
 
     setattr(sys.modules[__name__], database_name, DatabaseClass)
 
+# add databases from ~/.pyannote/protocols.yml if it exists
```
Comment (hbredin) [inline pyannote/database/__init__.py]: I'd rather use `custom.yml` than `protocols.yml`.

File: pyannote/database/filelist.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+def subset_iter(database_name, path_mdtm=None, path_uem=None):
+    """This function will become a xxx_iter method of a protocol.
+
+    Parameters
+    ----------
+    database_name : `str`
+        Database name.
+    path_mdtm : `Path`, optional
+        Path to MDTM file.
```
Comment (hbredin) [inline pyannote/database/filelist.py]: I think we should use `rttm` instead which seems to be more used than `mdtm`.
File format is not that different: https://github.com/nryant/dscore#rttm

File: pyannote/database/filelist.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+from pyannote.core import Segment, Timeline, Annotation
+from .database import Database
+from . import protocol as Protocol
+import functools
+from pathlib import Path
+import yaml
+import pandas as pd
+
+
+def subset_iter(database_name, path_mdtm=None, path_uem=None):
```
Comment (hbredin) [inline pyannote/database/filelist.py]: We should ask for a mandatory `path_lst` argument as well that contains one uri per line.
This will add support for files with no annotation (which would otherwise be ignored).

`path_rttm` and `path_uem` can then become really optional.
When `path_rttm` is missing, we don't add the "annotation" key.
When `path_uem` is missing, we don't add the "annotated" key.

File: pyannote/database/filelist.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        Path to UEM file.
+
+    Yields
+    ------
+    current_file : `dict`
+        Dictionary that provides information about a file.
+        This includes (at least) a 'uri' key and a 'database' keys.
+
+    """
+
```
Comment (hbredin) [inline pyannote/database/filelist.py]: Start by gathering uris from `path_lst` and then only iterate over the corresponding annotation groups and annotated groups (if they exist).

File: pyannote/database/filelist.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        yield {'uri': uri,
+               'database': database_name,
+               'annotation': reference,
+               'annotated': uem}
+
+
+def add_filelist_databases(databases={}, tasks={}, config_path=None):
+    """
+
+    :param databases: This dictionary cotains one database class per entry.
```
Comment (hbredin) [inline pyannote/database/filelist.py]: I tend to use docstrings with `numpydoc` style in the pyannote ecosystem.
Would be nice to keep using this...

**Comment by pkorshunov** [comment]
> Thanks @pkorshunov.
> 
> Could you please give me permission to push to your repo so that we can work together on this?
> See https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

The edit permissions for maintainers are already enabled, so you should be able to push here.

File: pyannote/database/filelist.py (hunk 3 above)
Comment (pkorshunov) [reply pyannote/database/filelist.py]: We could allow both formats and parse them based on the file extension.

File: pyannote/database/filelist.py (hunk 4 above)
Comment (pkorshunov) [reply pyannote/database/filelist.py]: I think `path_mdtm` has to be changed to a more generic `path_annotations` and we can allow it to be either in `mdtm`, `rttm`, or a simple `list` format (this one with extension `.lst`). So, we parse it according to its extension. So there will be only 2 path parameters and path_uem can be optional.

But I wonder, will pyannote work correctly if `uri` is yielded with both annotations and annotated keys empty? I assumed a pyannote database has to have some sort of annotations, even if the timeline covers the whole file.

Comment (hbredin) [reply pyannote/database/filelist.py]: The problem with allowing to pass `mdtm` but not `lst` is that there is no way to know that a file with no annotation exists (because it does not appear in the `mdtm` file).

Training SAD, SCD or EMB will fail when dictionary as no 'annotation' key. 
However, prediction will still work... and MUST work because for evaluation campaigns, we don't have access to annotations on the test set.

File: pyannote/database/filelist.py (hunk 3 above)
Comment (hbredin) [reply pyannote/database/filelist.py]: We could. I am just worried that this will get out of control...

I used to maintain [`pyannote.parser`](https://github.com/pyannote/pyannote-parser) for this very purpose but plan to drop support for it and only support a limited set of file formats...

File: pyannote/database/filelist.py (hunk 4 above)
Comment (pkorshunov) [reply pyannote/database/filelist.py]: Ah, you want to allow the same subset (train, development, or test) with mixed audio files: some with annotations and some without? Can it really happen in practice? I thought that one subset is always of a certain type. If you have a training set in a database, it would either have all utterances annotated or not, but it would never have half annotated and half not.

Comment (hbredin) [reply pyannote/database/filelist.py]: You are right: within a specific subset, it will be either 100% with annotation or 100% without annotation.

But it may happen that a 100% annotated file actually contains no annotation at all while still being correctly annotated (= does not contain any speech...).

Comment (hbredin) [reply pyannote/database/filelist.py]: This is because `mdtm` files don't provide "non-speech" labels...

Comment (pkorshunov) [reply pyannote/database/filelist.py]: Then, we still don't have a situation when the subset would have both files: the one with annotations in `rttm` (or `mdtm`) format and one a simple list file with just a single `uri` per line. Since it is either-or situation, I would think it's better to have just one file path parameter, so avoid confusion with all those different file paths. Three paths is a bit much.

Comment (hbredin) [reply pyannote/database/filelist.py]: Yes. we do. Say we have two files (`file1` and `file2`) in the training set. 
* `file1` is a *regular* file with some speech turns.
* `file2` is a file with only non-speech.

If you only provide the `mdtm`, then you will only iterate on `file1` (because the `mdtm` won't refer to `file2` at all).
This means that all the non-speech data from `file2` won't be used when training speech activity detection. We don't want that, right?

Comment (pkorshunov) [reply pyannote/database/filelist.py]: Ok, I didn't know such databases exist

**Comment by hbredin** [comment]
I just pushed a few commits to this pull-request.
I am now happy with the way it looks and will merge it into develop.

Be careful, I changed `custom.yml` format slightly: it no longer expects the root level that you had before (this is in preparation of a future merge of all ~/.pyannote/*yml` files).

**Comment by pkorshunov** [comment]
Ehh, I just added the support for a set of RTTM or MDTM files in a directory. Let me see if I can add it in a cleaner way since you have re-written almost everything

**Comment by hbredin** [comment]
Oooooh. Sorry :-/

**Comment by pkorshunov** [comment]
no problems, I should have checked your commits before pushing.

**Comment by hbredin** [comment]
> `pyannote.database` now has three different configuration files:
> 
> * `~/.pyannote/db.yml` is used by `FileFinder` to find files.
> * `~/.pyannote/meta.yml` is used to define meta-protocol
> * `~/.pyannote/protocols.yml` is used for this generic database interface.
> 
> I think this is getting too complex. We should merge everything into a unique `~/.pyannote/database.yml` whose structure has yet to be defined...
> 
> We could do that after this new feature is added, though...

FYI. I am now working on merging those three configuration files into just one.
